### DISCUSSION TITLE
Zed AI: OpenRouter Support

### DISCUSSION THREAD

**Discussion by daun-io** [zed-industries/zed#16576] [Feature Requests]
Hi, I'm from OpenRouter and a fan of Zed and would love to see OpenRouter integration as a new [assistant provider](https://zed.dev/docs/assistant/configuration#configuring-the-assistant).
OpenRouter supports dozens of the best coding models and [hundreds of different model](https://openrouter.ai/models)s with one API key and an OpenAI-style schema.
I saw that some Zed community users have requested this feature [1](https://github.com/zed-industries/zed/issues/11984) [2](https://github.com/zed-industries/zed/issues/14963).
Is there anything I can do to help? Thanks! 💙

**Comment by deanrie** [comment]
It would be nice to have this support.

**Comment by mintisan** [comment]
It would be nice to have this support.

**Comment by minato7namikazi** [comment]
it would be awesome  : )

**Comment by tm17-abcgen** [comment]
Would love to have this too. ZedAI currently works with Anthropic only. Since OpenRouter has access to Anthropic, i hope that ZedAI can be used with OpenRouter too, to make use of the prompt caching also!

**Comment by dkphhh** [comment]
we need this！

**Comment by hrefhref** [comment]
isn't this already supported : https://github.com/zed-industries/zed/pull/13276

**Comment by drivencompassion** [comment]
Awaiting OpenRouter [config](https://zed.dev/docs/assistant/configuration)?

**Comment by gliptic** [comment]
OpenRouter's API is OpenAI compatible, so just e.g.:
```
  "language_models": {
    "openai": {
      "version": "1",
      "api_url": "https://openrouter.ai/api/v1",
      "available_models": [
        {
          "display_name": "Qwen2.5-72B",
          "name": "qwen/qwen-2.5-72b-instruct",
          "max_tokens": 131072
        }
      ]
...
```

**Comment by drivencompassion** [comment]
Is there an example of configuring multiple models like openai/chatgpt-4o-latest, o1-mini & o1-preview?

**Reply by drivencompassion** [reply]
<img width="369" alt="Screenshot 2024-09-25 at 12 08 07 PM" src="https://github.com/user-attachments/assets/9fd170df-db71-4523-b03d-3a8f8c9e1faa">

**Reply by drivencompassion** [reply]
![image](https://github.com/user-attachments/assets/a79867d0-7525-4ab4-b2b4-e467718ab9cb)

**Comment by drivencompassion** [comment]
In addition to the LLM providers (e.g., the OpenAI API Key option), may there be an option for OpenRouter?

Run assistant (shift+command+p): show configuration

**Reply by bfulop** [reply]
I just had the same issue:

I had to add my **Openrouter API key** in the `export OPENAI_API_KEY =` env var and restart zed

**Reply by drivencompassion** [reply]
Where else is [env var](https://zed.dev/docs/assistant/configuration) stored? I get an `invalid settings` error when trying to add `"export OPENAI_API_KEY": ""` to `settings.json`.

**Reply by bfulop** [reply]
In my case it's in `.zshrc` file in my home directory.

**Reply by bfulop** [reply]
If you don't have this set in you shell, then you can set it via the `assistant: show configuration` and replace the OpenAI key with your openrouter key. In my case because it's set via the shell I cannot edit it.

![BalintFulop 2024-09-25--20-14-02@2x](https://github.com/user-attachments/assets/a3542ad9-8aee-4fbd-b5d7-e6f0ecd4bd23)

**Comment by azdanov** [comment]
<img width="1552" alt="Screenshot 2024-10-29 at 08 32 43" src="https://github.com/user-attachments/assets/3c5eacbb-9706-4ba2-89ea-81404a10cffb">

Does it work for anyone else?

**Reply by knealking** [reply]
It worked for me. But I had to restart the app.

**Comment by phil65** [comment]
Seems as if right now we cant "mix" multiple OpenAI(-compatible) APIs because they all would have to share the same API key. Would be great if Zed would get rid of these restrictions so that someone use multiple different OpenAi-compatible APIs simultaniously. :)

**Comment by netapy** [comment]
+1, is the Zed team on this ? It would massively open up the ability to use any models at token level pricing

**Comment by TYTheBeast** [comment]
Waiting on this as well, I just discovered Zed and this missing integration is a killer, I have an OpenRouter API key but no Anthropic API key.

**Comment by pconerly** [comment]
+1 for this. I want to throw deepseek-r1 at my problems!

**Reply by bfulop** [reply]
I'm using this. Not perfect (it overrides openai) but works:

```json
  "language_models": {
    "openai": {
      "api_url": "https://openrouter.ai/api/v1",
      "available_models": [
        {
          "name": "deepseek/deepseek-r1",
          "display_name": "DeepSeek R1",
          "max_tokens": 64000
        },
        {
          "name": "deepseek/deepseek-chat",
          "display_name": "DeepSeek V3",
          "max_tokens": 64000
        }
      ],
      "version": "1"
    }
  },

```

Then in my `.zshrc` I replaced the `OPENAI_API_KEY` with the openrouter key.

**Reply by pconerly** [reply]
Thanks! That worked. FYI @drivencompassion , i put these settings into `~/.config/zed/settings.json`, and then configured the 'openai' env variable via the Zed UI.

screenshot:
<img width="295" alt="Screenshot 2025-01-21 at 9 30 36 AM" src="https://github.com/user-attachments/assets/11c34bcb-622d-4a08-92ae-553acdb21389" />

**Comment by drivencompassion** [comment]
After configuring this method in settings, and in .zshrc  replacing key listed in OPENAI_API_KEY with the openrouter key, while keeping OPENAI_API_KEY=sk-or-v1- … of openrouter key, and testing a prompt, going to openrouter activity shows OpenAI: o1-mini model was selected, not DeepSeek R1

Anyone else try this and see different activity?

**Comment by drivencompassion** [comment]
~/.config/zed/settings.json, updated … how did you configure the 'openai' env variable via the Zed UI, as DeepSeek R1 is not an option under my config options like your screenshot.

----
On Jan 21, 2025 at 12:41:25 PM, Peter Conerly wrote:

Thanks! That worked. FYI @drivencompassion , i put these settings into ~/.config/zed/settings.json, and then configured the 'openai' env variable via the Zed UI.

screenshot:
Screenshot.2025-01-21.at.9.30.36.AM.png (view on web)

**Reply by pconerly** [reply]
In the context window (or the "llm chat window") there's a 3-dot button to open options, and one of those options is "Configure". I also gave Zed a restart but I don't know if that was necessary

![Screenshot 2025-01-21 at 10 38 43 AM](https://github.com/user-attachments/assets/dd23b922-e5fb-492e-b242-e531b3f2de70)

**Comment by drivencompassion** [comment]
Tried rebooting and restarting app, but 3-dot options does not have an option for DeepSeek R1, although settings does have:

  "language_models": {
  "openai": {
    "api_url": "https://openrouter.ai/api/v1",
    "available_models": [
      {
        "name": "deepseek/deepseek-r1",
        "display_name": "DeepSeek R1",
        "max_tokens": 64000
      },
      {
        "name": "deepseek/deepseek-chat",
        "display_name": "DeepSeek V3",
        "max_tokens": 64000
      }
    ],
    "version": "1"
  }
}

----
On Jan 21, 2025 at 1:40:15 PM, Peter Conerly wrote:

In the context window there's a 3-dot button to open options, and one of those options is "Configure". I also gave Zed a restart but I don't know if that was necessary
Screenshot.2025-01-21.at.10.38.43.AM.png (view on web)

**Comment by amin-abouee** [comment]
I've tried both OpenRouter and the experimental Gemini model, but neither worked for me.

"language_models": {
    "openai": {
      "api_url": "https://openrouter.ai/api/v1",
      "available_models": [
        {
          "name": "deepseek/deepseek-r1",
          "display_name": "DeepSeek R1",
          "max_tokens": 64000
        },
        {
          "name": "deepseek/deepseek-chat",
          "display_name": "DeepSeek V3",
          "max_tokens": 64000
        }
      ],
      "version": "1"
    },
    "google": {
      "available_models": [
        {
          "name": "gemini-exp-1206",
          "display_name": "Gemini 1206",
          "max_tokens": 1000000
        },
        {
          "name": "gemini-2.0-flash-exp",
          "display_name": "Gemini 2.0 Flash",
          "max_tokens": 1000000
        }
      ]
    }
  },

**Comment by realmorrisliu** [comment]
Hey folks, just wanted to chime in — I recently put together a Rust SDK for OpenRouter, and I’ve also started working on adding OpenRouter support to Zed. I’ve been using OpenRouter quite a bit, and figured it’d be great if Zed could support it out of the box.

It seems no one else is actively working on this, so I’ve decided to take it on. Happy to collaborate if anyone’s interested!

**Reply by RunLittleTurtle** [reply]
Wow that would be amazing, keep us updated.

**Reply by miguilimzero** [reply]
Thank you for the effort mate!

**Reply by realmorrisliu** [reply]
Latest update: I’m currently modifying parts of the OpenRouter SDK to implement the traits expected by Zed’s AI provider. Hopefully, I’ll be able to start testing it on my Mac this week—specifically, seeing OpenRouter show up in Zed’s UI. I’m still figuring out the best way to include OpenRouter’s configuration within Zed’s config file. If all goes well, it should be usable in a few weeks. I’ve had quite a bit of other work recently, so progress has been a bit slower than I’d like.

**Reply by realmorrisliu** [reply]
I’ve fixed some type issues, and the basic OpenRouter UI is now rendering correctly in Zed. Next, I’ll be adding more functionality and refining the interactions.
<img width="1136" alt="image" src="https://github.com/user-attachments/assets/4419233c-77d3-4296-ab5c-b133b5b86fb4" />
<img width="1136" alt="image" src="https://github.com/user-attachments/assets/91e74f49-418b-4ae1-9a0a-21c3829d9deb" />
The model name isn’t displaying correctly at the moment—I’ll look into what’s going wrong.
<img width="1136" alt="image" src="https://github.com/user-attachments/assets/ef3730bd-3209-4008-8f5a-fa45931bfc22" />

**Reply by aaditagrawal** [reply]
Can this be published as an extension of some sort? That'd be incredible.

**Reply by realmorrisliu** [reply]
> Can this be published as an extension of some sort? That'd be incredible.

Yeah, that does sound like a great idea! I haven’t looked into how Zed extensions work yet—still focusing on getting OpenRouter up and running first. Once it’s in a usable state, I’d love to explore that path.

**Reply by AvMavs** [reply]
<img width="638" alt="image" src="https://github.com/user-attachments/assets/7139c347-6d64-4730-8e81-09988d5f8a95" />

In terms of UX, something like the amazon bedrock selector could be cool, where we have the ability to manually add the model address/name and the API key alongside it.

**Reply by realmorrisliu** [reply]
> In terms of UX, something like the amazon bedrock selector could be cool, where we have the ability to manually add the model address/name and the API key alongside it.

Yeah, I’ve been thinking about adding support for custom model input—since OpenRouter supports so many models, I plan to offer a few common ones by default, but still leave room for users to add their own by name.

As for what you mentioned about the API key—are you thinking of supporting multiple API keys? Would love to hear more about your use case!

**Reply by aaditagrawal** [reply]
Oh, not that; I was wondering if we could have multiple fields, that's all. The other LLM configurations only have a slot for the API keys, but for OpenRouter, we can pass multiple arguments, and that helps quite a bit!

**Reply by sysradium** [reply]
Will the model selection not explode if you fetch all the models from openrouter? :)

**Reply by realmorrisliu** [reply]
> Will the model selection not explode if you fetch all the models from openrouter? :)

At first, I was planning to fetch all models from OpenRouter, but later decided to offer a smaller set of commonly used programming models by default, with the option for users to manually add any others. Feels like a good balance this way!

**Reply by sysradium** [reply]
@realmorrisliu I agree, makes a lot of sense.

**Reply by t18n** [reply]
You can consider to implement OWU approach, which allow user to add models from OpenRouter using the model id.
![image](https://github.com/user-attachments/assets/c398b430-a1e0-4e54-88ac-ae1f62305355)

**Reply by realmorrisliu** [reply]
> You can consider to implement OWU approach, which allow user to add models from OpenRouter using the model id. 

Looks great!

**Reply by sambarnes** [reply]
oooh hell yeah! stoked to try this out

**Reply by sambarnes** [reply]
we also recently added support for sending a `category=programming` query param to the models endpoint, which returns the top 20 coding models, in case that's helpful at all here

 https://openrouter.ai/api/v1/models?category=programming

**Reply by imumesh18** [reply]
@sambarnes, any chance we could get the tool_use supported and category in the response of the list models itself. Because currently i am making like three separate api calls for getting the data, one for all models, one for tool use and then one for programming specific. It could be much simpler if we just get additional parameters in the response itself.

**Reply by sambarnes** [reply]
@imumesh18 we're not sure about including category data there yet, but can def add the supported params in the response.  I'll let you know when it's complete, should be able to get to it in the next few days

**Reply by realmorrisliu** [reply]
> we also recently added support for sending a `category=programming` query param to the models endpoint, which returns the top 20 coding models, in case that's helpful at all here
> 
> https://openrouter.ai/api/v1/models?category=programming

That’s super helpful! I’ll start by adding support for this parameter in openrouter-rs, and then see how to integrate it into Zed.

**Reply by imumesh18** [reply]
Thanks @sambarnes that's really helpful.

**Reply by yeferson59** [reply]
Hi! How is the integration with OpenRouter going? Have you had any problems or difficulties with it?

**Reply by realmorrisliu** [reply]
> Hi! How is the integration with OpenRouter going? Have you had any problems or difficulties with it?

Still working on it! Things have been a bit busy lately so progress is a little slower than expected. I’m aiming to get a minimal working version out first, and then build on it with more features.

**Reply by imumesh18** [reply]
I have a open PR: https://github.com/zed-industries/zed/pull/29496 for openrouter integration. In case anyone wants to try and provide the feedback. Please feel free to do so by checking out the branch. There are still few open questions in terms of ux and what is acceptable from zed team. As they have been busy with agent panel they haven't reviewed or provided any feedback. Hopefully someone soon will take a look.

**Reply by realmorrisliu** [reply]
Looks like @imumesh18 has already submitted a solid implementation, so I’ll step back from developing my own version. The PR looks great—hopefully it gets merged soon!

**Reply by sambarnes** [reply]
@imumesh18 ay sorry for not letting you know sooner, but we merged this last week:
https://openrouter.ai/docs/api-reference/list-available-models

main caveat for why we hadnt added this sooner:

> Returns a list of models available through the API. Note: `supported_parameters` is a union of all parameters supported by all providers for this model. There may not be a single provider which offers all of the listed parameters for a model.

so like for a model with 10 providers, if you needed literally every parameter in the returned union and then use the [provider.require_parameters](https://openrouter.ai/docs/features/provider-routing) flag, it's possible that it'd 404

much less of an issue for the flagship models tho, just thought its worth calling out

**Reply by imumesh18** [reply]
Hey @sambarnes , needed your help verify something we are sending below headers but on the openrouter dashboard it's not getting tagged properly. It is getting tagged as unknown. Can you point out are we doing something wrong here.

```
        .header("HTTP-Referer", "zed.dev")
        .header("X-Title", "Zed Editor");
```

**Reply by imumesh18** [reply]
Never mind fixed it. It seems that openrouter requires url in this format: `https://zed.dev`.

**Reply by sambarnes** [reply]
@imumesh18 yo just waking up, best guess is put https:// in front of the url

but assuming y'all tried that, once I'm at my computer I can try to take a closer look and repro -- there's some debt in this part of our system so could be on us here

**Reply by imumesh18** [reply]
Hey thanks @sambarnes for responding you are correct need to prefix https:// before the url for it to get tagged.

**Reply by sambarnes** [reply]
right on! can't wait to see zed climb the leaderboard (:

**Reply by imumesh18** [reply]
Hey @sambarnes so openrouter went live on zed couple weeks back since then it has been climbing the ladder. Thinking and images is also going live soon. But one of the major complaints from users has been about the number of models being overwhelming high and model filtering. Just wondering do you guys have any plans to introduce authenticated model api which will return limited number of models enabled by users from openrouter dashboard?

**Reply by imumesh18** [reply]
This can be done from zed also but quite a lot of hoops to get ux approved. So just wanted to know if there exists or will exists a api which can solve this without much changes from zed side.

**Comment by jedisct1** [comment]
Pretty crazy that Openrouter is still not supported.

**Reply by imumesh18** [reply]
Hey there is open pr: https://github.com/zed-industries/zed/pull/29496. The best way you can speed up getting this landed is by testing it and reporting issues/feedback.

**Comment by mishaal79** [comment]
hey any update on support for openrouter ?
can we build this and make a pull request ?

**Reply by imumesh18** [reply]
Hey there is open pr: https://github.com/zed-industries/zed/pull/29496. The best way you can speed up getting this landed is by testing it and reporting issues/feedback.

**Comment by vavans** [comment]
The configuration 
```json
"language_models": {
    "openai": {
      "api_url": "https://openrouter.ai/api/v1",
      "version": "1"
    }
```
is working for some open ai models (gpt-4.1 for instance) but i got an error for o4-mini:
![image](https://github.com/user-attachments/assets/0ee3e573-259a-499b-9a59-96df6bad83a0)

**Comment by canadaduane** [comment]
Another option is to create a local gateway that connects to openrouter: https://github.com/fabiojbg/LLMApiGateway

You can create specific routing requirements this way, such as provider or fallbacks based on rules.

I know it isn't ideal to have a router behind a router, but it is a decent workaround while waiting for the OpenRouter-related PR to land.

**Comment by huedaya** [comment]
this worked for me, but sometimes got `Error interacting with language model, Failed to connect to OpenAI API: Provider returned error` and have no idea why.

Any ideas how to debug whats error being returned?

```json5
{
  "language_models": {
    "openai": {
      "api_url": "https://openrouter.ai/api/v1",
      "available_models": [
        {
          "name": "google/gemini-2.5-pro-preview",
          "display_name": "2.5-pro-preview",
          "max_tokens": 1048576
        },
      ],
      "version": "1"
    },
  },
}
```

**Comment by gengurke** [comment]
Support for OpenRouter has been merged to main! https://github.com/zed-industries/zed/pull/29496 Let's hope it lands in preview soon!

**Reply by yeferson59** [reply]
Yeah, yesterday it was merged to main

**Reply by yeferson59** [reply]
The preview with the support is already out!!

**Reply by vavans** [reply]
Nice !!

**Reply by josephleblanc** [reply]
Yep, this is what I needed to have before using zed.

**Comment by grantstephens** [comment]
This discussion should be closed. Support is available in the latest release.

**Comment by dinocosta** [comment]
Hey 👋 

Looks like we forgot to close this discussion, given that OpenRouter support has already been shipped! Documentation is available at [Zed – LLM Providers: OpenRouter](https://zed.dev/docs/ai/llm-providers#openrouter) .

If you happen to find any issues while using it, feel free to open a bug report so we can take a look. Thanks! 🙂
